backend/remote-state/artifactory/client.go

```diff
@@ -48,16 +79,88 @@ func (c *ArtifactoryClient) Get() (*remote.Payload, error) {
 }
 
 func (c *ArtifactoryClient) Put(data []byte) error {
+	var err error
+	logStart("Put")
+	defer logResult("Put", &err)
+
 	p := fmt.Sprintf("%s/%s/%s", c.repo, c.subpath, ARTIF_TFSTATE_NAME)
-	if _, err := c.nativeClient.Put(p, string(data), make(map[string]string)); err == nil {
+	if _, err = c.nativeClient.Put(p, string(data), make(map[string]string)); err == nil {
 		return nil
 	} else {
 		return fmt.Errorf("Failed to upload state: %v", err)
 	}
 }
 
 func (c *ArtifactoryClient) Delete() error {
+	var err error
+	logStart("Delete")
+	defer logResult("Delete", &err)
 	p := fmt.Sprintf("%s/%s/%s", c.repo, c.subpath, ARTIF_TFSTATE_NAME)
-	err := c.nativeClient.Delete(p)
+	err = c.nativeClient.Delete(p)
+	return err
+}
+
+func (c *ArtifactoryClient) Lock(info *state.LockInfo) (string, error) {
+	if c.lockUrl == "" {
+		return "", nil
+	}
+	c.lockID = ""
+	var err error
+	var output []byte
+	logStart("Lock")
+	defer logResult("Lock", &err)
+	rand.Seed(time.Now().UnixNano())
+	info.ID = fmt.Sprintf("%s-%d", info.ID, rand.Intn(100000000))
+	jsonLockInfo := info.Marshal()
+	p := fmt.Sprintf("%s/%s/%s", c.lockRepo, c.lockSubpath, ARTIF_TFLOCK_NAME)
+	if _, err = c.lockNativeClient.Put(p, string(jsonLockInfo), make(map[string]string)); err == nil {
+		if c.lockReadbackWait <= 0 {
+			c.lockID = info.ID
+			c.jsonLockInfo = jsonLockInfo
+			return info.ID, nil
+		}
+		time.Sleep(time.Duration(c.lockReadbackWait) * time.Millisecond)
+		// readback and compare with original info
+		output, err = c.lockNativeClient.Get(p, make(map[string]string))
```

Comment:
What is the consistency model of the artifactory storage? If there is any possibility that reading the lock data here could fail without a delay, then this is absolutely not suitable for a lock implementation. The locks must be strongly consistent, and hence checking this data a second time should never be required.